lib/phlex/sgml.rb

```diff
@@ -433,5 +449,78 @@ def __attributes__(attributes, buffer = +"")
 
 			buffer
 		end
+
+		# @api private
+		def __classes__(c)
+			case c
+			when String
+				c
+			when Symbol
+				c.name
+			when Array, Set
+				c.filter_map { |c| __classes__(c) }.join(" ")
+			when Hash
+				c.filter_map { |c, add|
+					next unless add
+					case c
+						when String then c
+						when Symbol then c.name.tr("_", "-").delete_suffix("?")
+						else raise ArgumentError, "Class keys should be Strings or Symbols."
+					end
+				}.join(" ")
+			when nil, false
+				nil
+			else
+				if c.respond_to?(:to_phlex_attribute_value)
+					c.to_phlex_attribute_value
+				elsif c.respond_to?(:to_str)
+					c.to_str
+				else
+					c.to_s
+				end
+			end
+		end
+
+		# @api private
+		def __styles__(s)
+			style = case s
+			when String
+				s
+			when Symbol
+				s.name
+			when Integer, Float
+				s.to_s
+			when Array, Set
+				s.filter_map { |s| __styles__(s) }.join
+			when Hash
+				buffer = +""
+				s.each do |k, v|
+					prop = case k
+						when String then k
+						when Symbol then k.name.tr("_", "-")
+						else raise ArgumentError, "Style keys should be Strings or Symbols."
+					end
+
+					value = __styles__(v)
+
+					if value
+						buffer << prop << ":" << value
+					end
+				end
+				buffer
+			when nil, false
+				return nil
+			else
+				if s.respond_to?(:to_phlex_attribute_value)
+					s.to_phlex_attribute_value
+				elsif s.respond_to?(:to_str)
+					s.to_str
+				else
+					s.to_s
+				end
+			end
```

Comment:
This chunk is repeated 3x now. I wonder if it might be worth extracting a method for it, but that could impact performance.

Reply:
To be clear, I’m fine with this repetition as long as it’s worth it for the perf benefits.

Reply:
I assumed inlining would benefit performance, but I agree we should benchmark it :+1:

The introduction of YJIT challenges some of the usual outcomes, perhaps extracting to a method no longer has a significant impact.

Reply:
I'm also curious to know if extracting it does help or not. I could see that that would make it a more frequented code path so it might get JITed sooner. I don't think we need to hold this up for that experiment though.